apps/src/javalab/Playground.js

```diff
@@ -128,38 +128,59 @@ export default class Playground {
   }
 
   addTextItem(itemData) {
-    if (this.isGameOver) {
-      // can't add new items if the game is over
+    if (this.isGameOver || this.itemExists(itemData)) {
+      // can't add new items if the game is over or if the item already exists
       return;
     }
+
+    const textData = {...itemData};
+    delete textData.id;
+    textData.type = PlaygroundItemType.TEXT;
+
+    this.addPlaygroundItem(itemData.id, textData);
   }
 
   removeItem(itemData) {
     if (this.isGameOver) {
       // can't remove items if game is over
       return;
     }
-    if (this.imageItemExists(itemData)) {
+    if (this.itemExists(itemData)) {
       this.removePlaygroundItem(itemData.id);
     }
-    // TODO: handle text deletion
   }
 
   changeItem(itemData) {
     if (this.isGameOver) {
       // can't change items if game is over
       return;
     }
-    if (this.imageItemExists(itemData)) {
-      const newImageData = {...itemData};
+
+    const newItemData = this.getNewItemData(itemData);
+    this.changePlaygroundItem(itemData.id, newItemData);
+  }
+
+  getNewItemData(itemData) {
+    // We do not include the ID as part of each item's data.
+    // The ID serves as the key referencing an object that contains the item's contents.
+    const newItemData = {...itemData};
+    delete newItemData.id;
+
+    if (this.getItem(itemData.id).type === PlaygroundItemType.IMAGE) {
       if (itemData.filename) {
-        newImageData.fileUrl = this.getUrl(itemData.filename);
+        newItemData.fileUrl = this.getUrl(itemData.filename);
         // we don't need to pass filename as imageData
-        delete newImageData.filename;
+        delete newItemData.filename;
       }
-      this.changePlaygroundItem(itemData.id, newImageData);
+      return newItemData;
+    }
+
+    if (this.getItem(itemData.id).type === PlaygroundItemType.IMAGE) {
+      this.changePlaygroundItem(itemData.id, newItemData);
```

Comment:
should this be here? Looks like we already call `this.changePlaygroundItem` in `changeItem`

Reply:
It should not, missed removing on refactor. Thanks!

Reply:
Also, used the wrong constant 🤦 . Basically, we don't need to make any changes to `itemData` for text.